spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/firestore/GcpFirestoreEmulatorAutoConfigurationTests.java

```diff
@@ -0,0 +1,96 @@
+ */
+
+package com.google.cloud.spring.autoconfigure.firestore;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.api.gax.core.NoCredentialsProvider;
+import com.google.api.gax.grpc.InstantiatingGrpcChannelProvider;
+import com.google.auth.Credentials;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.FirestoreOptions;
+import com.google.cloud.spring.autoconfigure.core.GcpContextAutoConfiguration;
+import com.google.cloud.spring.core.DefaultCredentialsProvider;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.context.annotation.Bean;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests for the Firestore emulator config.
+ *
+ * @author Gary Wong
+ */
+class GcpFirestoreEmulatorAutoConfigurationTests {
+
+	private final ApplicationContextRunner contextRunner = new ApplicationContextRunner()
+			.withConfiguration(AutoConfigurations.of(
+					GcpContextAutoConfiguration.class,
+					GcpFirestoreEmulatorAutoConfiguration.class,
+					GcpFirestoreAutoConfiguration.class))
+			.withUserConfiguration(GcpFirestoreEmulatorAutoConfigurationTests.TestConfiguration.class);
+
+	@Test
+	void testEmulatorEnabledConfig() {
+		this.contextRunner.withPropertyValues(
+				"spring.cloud.gcp.firestore.projectId=test-project",
+				"spring.cloud.gcp.firestore.emulator.enabled=true",
+				"spring.cloud.gcp.firestore.host-port=localhost:8080"
+		).run(context -> {
+			CredentialsProvider defaultCredentialsProvider = context.getBean(CredentialsProvider.class);
+			assertThat(defaultCredentialsProvider).isNotInstanceOf(NoCredentialsProvider.class);
+
+			GcpFirestoreAutoConfiguration firestoreAutoConfiguration = context.getBean(GcpFirestoreAutoConfiguration.class);
+			assertThat(firestoreAutoConfiguration.getCredentialsProvider()).isInstanceOf(NoCredentialsProvider.class);
+
+			FirestoreOptions datastoreOptions = context.getBean(Firestore.class).getOptions();
+			assertThat(datastoreOptions.getProjectId()).isEqualTo("test-project");
+
+			InstantiatingGrpcChannelProvider channelProvider = (InstantiatingGrpcChannelProvider) datastoreOptions.getTransportChannelProvider();
+			assertThat(channelProvider.getEndpoint()).isEqualTo("localhost:8080");
+		});
+	}
+
+	@Test
+	void testDefaultConfig() {
+		this.contextRunner.withPropertyValues(
+				"spring.cloud.gcp.firestore.projectId=test-project"
+		).run(context -> {
+			GcpFirestoreProperties gcpFirestoreProperties = context.getBean(GcpFirestoreProperties.class);
+			assertThat(gcpFirestoreProperties.getEmulator().isEnabled()).isFalse();
+
+			CredentialsProvider defaultCredentialsProvider = context.getBean(CredentialsProvider.class);
+			assertThat(defaultCredentialsProvider).isNotInstanceOf(DefaultCredentialsProvider.class);
```

Comment:
Two minor testing things:
1) It would be good to validate that the mocked `CredentialsProvider` is returned.  One way to do that is to use `.withBean(() -> /* handle to mock */ )` instead of the current `.withUserConfiguration(GcpFirestoreEmulatorAutoConfigurationTests.TestConfiguration.class)`.
2) Could you also test `firestoreAutoConfiguration.getCredentialsProvider()` here?

Reply:
@elefeint thanks for the review! pushed new commit, would you mind have a look and let me know what you think ?